### PR TITLE
Support cipher suite selection

### LINKED ISSUE #4: Support cipher suite selection
OpenSSL's the most flexible here, and SChannel is the least. The cross-platform solution will probably look something like an SChannel style enum of algorithms that will be unioned:

```rust
pub enum Algorithm {
    Sha256,
    Aes,
    Dhe,
    Rsa,
    ...
}
```

If you then call
```rust
client_builder.supported_algorithms(&[Algorithm::Rsa, Algorithm::Aes, Algorithm::Sha256, Algorithm::Dhe])
```
would mean all ciphers that use RSA or DHE for key exchange, AES as a bulk cipher, *and* SHA256 as a hash algorithm. That'd translate in SChannel to
```rust
cred_builder.supported_algorithms(&[Algorithm::Rsa, Algorithm::Aes, Algorithm::Sha256. Algorithm::Dhe])
```
and in OpenSSL to (note that we have to generate the cartesian product of key exchange, bulk cipher and hash algorithms)
```rust
ctx.set_cipher_suite("RSA+AES+SHA256:DHE+AES+SHA256:@STRENGTH")
```
and in Secure Transport to a somewhat complicated dance where we load supported ciphers and filter them through the provided algorithms:
```rust
let supported = ctx.supported_ciphers();
let enabled = supported.into_iter().filter(|c| suite_supported(c, algorithms)).collect();
ctx.set_enabled_ciphers(&enabled);
```

One interesting question is what to do if no algorithm in a category is specified (e.g. just `&[Algorithm::Aes, Algorithm::Dhe]` is passed). Should that mean that no suites match and no connections will work, or that *all* algorithms of that type can be used?

An alternative design would be to have separate enums for each algorithm type:
```
client_builder.supported_algorithms(&[KeyExchange::Rsa, KeyExchange::Dhe], &[BulkCipher::Aes], &[Hash::Sha256]);
```
It would be a bit more clear, but more verbose as well.

### PR DESCRIPTION
**Pull request by cyang1**
The exported API is implemented as an explicit whitelist with a consistent default across platforms based on the allowed ciphersuites in rust-openssl, which is itself based on the list from Python.

The implementation for each backend is roughly what you'd expect.
  - For OpenSSL, it constructs a cartesian product of the selected algorithms, joined together with `+`.
  - For Schannel, it just maps the chosen algorithms to the correct Schannel equivalent.
  - For the Security Framework, it doesn't look like there's any structure behind the ciphersuite constants or more user-friendly APIs for interacting with them, so I've combed through the ciphersuite list and manually compiled lists of cipher suites that use each algorithm.

SHA512 is not included as a possible hash algorithm because it appears to be only available in the Security Framework.

Closes #4.

### CODE REVIEW

File: src/imp/openssl.rs [hunk 1]
```diff
@@ -11,14 +11,93 @@ use self::openssl::ssl::{
...
+            TlsSignatureAlgorithm::Rsa => "aRSA",
+            TlsSignatureAlgorithm::__NonExhaustive => unreachable!(),
+        }),
+    );
+    cipher_suite_strings = cartesian_product(
+        cipher_suite_strings,
+        cipher_suites.bulk_encryption.iter().map(|alg| match alg {
+            TlsBulkEncryptionAlgorithm::Aes128 => "AES128",
+            TlsBulkEncryptionAlgorithm::Aes256 => "AES256",
+            TlsBulkEncryptionAlgorithm::Des => "DES",
```
Comment (sfackler) [inline src/imp/openssl.rs:71]: Do we really need to offer historical curiosities like DES and RC2?

File: src/imp/openssl.rs [hunk 2]
```diff
@@ -11,14 +11,93 @@ use self::openssl::ssl::{
...
+            TlsBulkEncryptionAlgorithm::Rc2 => "RC2",
+            TlsBulkEncryptionAlgorithm::Rc4 => "RC4",
+            TlsBulkEncryptionAlgorithm::TripleDes => "3DES",
```
Comment (sfackler) [inline src/imp/openssl.rs:74]: This should include ChaCha20 IMO.

File: src/imp/openssl.rs [hunk 3]
```diff
@@ -11,14 +11,93 @@ use self::openssl::ssl::{
...
+            TlsKeyExchangeAlgorithm::Ecdhe => "ECDHE",
+            TlsKeyExchangeAlgorithm::Rsa => "kRSA",
+            TlsKeyExchangeAlgorithm::__NonExhaustive => unreachable!(),
+        }]
+    }));
+
+    cipher_suite_strings = cartesian_product(
+        cipher_suite_strings,
+        cipher_suites.signature.iter().map(|alg| match alg {
+            TlsSignatureAlgorithm::Dss => "aDSS",
```
Comment (sfackler) [inline src/imp/openssl.rs:60]: Like below, do we really need DSS?

File: src/imp/security_framework.rs [hunk 4]
```diff
@@ -38,6 +43,485 @@ lazy_static! {
...
+    CipherSuite::TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_RSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_DH_DSS_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_DH_RSA_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_DHE_DSS_WITH_3DES_EDE_CBC_SHA,
+    CipherSuite::TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA,
+    // CipherSuite::TLS_PSK_WITH_3DES_EDE_CBC_SHA,
```
Comment (sfackler) [inline src/imp/security_framework.rs:56]: Why are some of these commented out?

File: src/imp/security_framework.rs [hunk 5]
```diff
@@ -38,6 +43,485 @@ lazy_static! {
...
+    }
+}
+
+fn expand_algorithms(cipher_suites: &CipherSuiteSet) -> Vec<CipherSuite> {
+    let first = cipher_suites
+        .key_exchange
+        .iter()
+        .flat_map(key_exchange_alg_to_cipher_suites)
+        .copied();
+    let rest: &[HashSet<_>] = &[
```
Comment (sfackler) [inline src/imp/security_framework.rs:499]: It might be a bit better to load the key exchange ciphers into a hashset and then iterate through the the signature, bulk_encryption, etc ciphers to progressively remove bad ciphers from the set, and then load the remainder into the returned vec.

File: src/test.rs [hunk 6]
```diff
@@ -417,4 +417,28 @@ mod tests {
 
         p!(j.join());
     }
+
+    #[test]
+    fn badssl_cipher_suites_no_rsa() {
+        let builder = p!(TlsConnector::builder().supported_cipher_suites(
+            // Oddly, on Windows, allowing RSA key exchange, but not RSA signature algorithms still
```
Comment (sfackler) [inline src/test.rs:424]: This is the kind of edge case I'm worried about in exposing this level of control over cipher suites.

File: src/imp/openssl.rs (hunk 1 above)
Comment (cyang1) [reply src/imp/openssl.rs:71]: I lean towards not being opinionated about what is available. I can remove them if that's your preference, but I can imagine that someone might have some use for them for connecting to old, insecure servers.

File: src/imp/openssl.rs (hunk 2 above)
Comment (cyang1) [reply src/imp/openssl.rs:74]: That would be nice, but AFAICT Schannel doesn't have any ChaCha20 ciphersuites, and I've just included the GCD of the available algorithms across the different backends.

File: src/imp/security_framework.rs (hunk 4 above)
Comment (cyang1) [reply src/imp/security_framework.rs:56]: All the commented out ones are PSK. I can remove them, they're mostly left in as I was checking my work while manually compiling the lists, and for anyone else checking it over to see that it wasn't forgotten, but explicitly excluded.

File: src/imp/security_framework.rs (hunk 5 above)
Comment (cyang1) [reply src/imp/security_framework.rs:499]: That's a good point, I can change this.

File: src/test.rs (hunk 6 above)
Comment (cyang1) [reply src/test.rs:424]: FWIW, RSA is the only one that's valid in two different parts of the ciphersuite specification, so it's the only one this can happen to. Did some searching and it seems like it might be an expected quirk of Schannel that `RSA_KEYX` implies `RSA_SIGN`, although all the discussions that I found were on tangential topics ([1], [2]).

[1]: https://stackoverflow.com/questions/45211765/cryptoapi-rsa-schannel-provider-error
[2]: https://stackoverflow.com/questions/3568695/ms-csp-difference-between-at-signature-and-rsa-key-sign-and-also-at-keyexchang

File: src/imp/openssl.rs (hunk 2 above)
Comment (sfackler) [reply src/imp/openssl.rs:74]: I think it's fine that schannel doesn't support it yet - many versions of OpenSSL don't either. As long as it's paired with another algorithm that is supported it'll work.

Comment (cyang1) [reply src/imp/openssl.rs:74]: Sure, what's the experience you had in mind, though, if the user chooses to only specify ChaCha20 on Windows? As currently written, nothing here returns a `Result`, and from some testing, it looks passing no bulk encryption algorithms to the `supported_algorithms` call in Schannel will result in Windows choosing some (most likely system-wide) defaults. 

On Windows 10 19h1 where I'm trying this out, it looks like it chose to allow AES-128, AES-256, and 3DES, which seems consistent with https://docs.microsoft.com/en-us/windows/win32/secauthn/tls-cipher-suites-in-windows-10-v1903, but probably not ideal as far as the behavior of this API.

Comment (sfackler) [reply src/imp/openssl.rs:74]: It would need to fail to handshake, yeah. If that's not easily doable though we can leave it out for now.

File: src/imp/security_framework.rs (hunk 5 above)
Comment (cyang1) [reply src/imp/security_framework.rs]: Made an attempt at this, although I'm not entirely convinced that it's better that it was before.

**Comment by rneswold** [comment]
Can this be merged? I'm getting flagged by my site security department because I'm advertising 3DES. It would be nice to be able to restrict the ciphers.

**Comment by k3d3** [comment]
I'm interested in this getting merged as well. For the insecure ciphers, perhaps it would be a good idea to call them `DANGER_DES` or something similar, to keep with the trend of prepending danger to things that most people would not want to do?

**Comment by NikosEfthias** [comment]
any updates on this ?
